core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMergeJoin.java

```diff
@@ -101,28 +103,55 @@ public static boolean isMergeJoinSupported(JoinRelType joinType) {
     // Required collation keys can be subset or superset of merge join keys.
     RelCollation collation = required.getTrait(RelCollationTraitDef.INSTANCE);
     List<Integer> reqKeys = RelCollations.ordinals(collation);
-    ImmutableBitSet reqKeySet = ImmutableBitSet.of(reqKeys);
+    // need to copy reqKeys because it is not sortable.
+    reqKeys = new ArrayList<>(reqKeys);
+    List<Integer> leftKeys = immutableIntListToList(joinInfo.leftKeys, 0);
+    List<Integer> rightKeys = immutableIntListToList(joinInfo.rightKeys,
+        left.getRowType().getFieldCount());
+    List<Integer> rightKeysNotShifted = immutableIntListToList(joinInfo.rightKeys, 0);
 
+    ImmutableBitSet reqKeySet = ImmutableBitSet.of(reqKeys);
     ImmutableBitSet leftKeySet = ImmutableBitSet.of(joinInfo.leftKeys);
     ImmutableBitSet rightKeySet = ImmutableBitSet.of(joinInfo.rightKeys)
         .shift(left.getRowType().getFieldCount());
 
     Map<Integer, Integer> keyMap = new HashMap<>();
-    final int keyCount = leftKeySet.cardinality();
+    final int keyCount = leftKeys.size();
```

Comment:
nice catch. in case the keys are `1,1,2,2`. IMHO, we shouldn't see duplicate join keys in physical merge/hash join operators. They should be optimized away, because that means the predicate is not pushed down at all.
like foo.a = bar.b and foo.a=bar.c. The predicate bar.b=bar.c should be pushed down for table bar.